Dev mount fix

jdejoya17 opened the pull request:
- issue: neither `mount_dir` or `mount_over_dir` is given, this will
result in `run_command` failing.
- solution: the internal function `is_fspath_mounted` already has
built in logic to fail early given the above scenario, so instead of
having `run_command` invoked and then fail, then we might as
well fail early

- issue: when unmount is successful, module does not report
`changed`
- solution: since the module did actually unmount a filesystem
successfully, then `changed` must be reported in ansible

- issue: in the case where `mount_over_dir` and `mount_dir`
are both given, `mount_dir` act as the backing device and 
`mount_over_dir` is where the mount point of the filesystem.
(1) the first issue is in previous logic does not check `mount_over_dir` 
if it is already mounted in this case;
(2) the second issue, appropriate reporting message is not selected in
the case where `mount_dir`/`mount_over_dir` is already mounted/unmounted
- solution: make sure to check for `mount_over_dir` in the above message
and make sure to select the appropriate reporting message

- issue: the feature to allow mounting by mount group (mount all
filesystems on the same mount group) is not idempotent
- solution: make sure that to report those filesystems in the mount
group that are already mounted and those that are mounted successfully.
if ALL the filesystems in the mount group are already mounted then
report `not changed`

- issue: report says failed during unmount when unmounting by 
mount group, all remote filesystems, or by node, when actually
there was nothing to unmount, because those filesystems are
already unmounted
- solution: make sure to report nothing changed instead of failing

- issue: previously, listing all mounted filesystems is a supported
feature of mount module. however, it wasn't quite clear on the
usage, (you had to not define any parameters for it to do this
feature)
- solution: improve the feature by making the usage more clear.
To do this, the new choice for `state` parameter called `show`.
If the state is `show` all the other parameters are ignored.

- added new unit tests for mount module
- added new examples and updated the existing examples in
the mount module

jdejoya17 committed:
- if neither mount_dir or mount_over_dir is given
then fail for mount state, while fail early for
umount state instead of letting run_command fail
early
- this check is handled by the is_fspath_mounted

jdejoya17 committed:
- make sure that when unmounting is successful, then
mount module should report changed

raeyacld commented on plugins/modules/mount.py:
```diff
@@ -113,10 +113,13 @@
-- name: Mount filesystems provided by a node"
+- name: |
+    Mount a filesytem, /tmp/servnfs provided by
+    a node on the mount point /tmp/clientnfs"
```
Is there a particular reason of changing the mount directory from /mnt to /tmp? /mnt is AIX default mount and usually /tmp has small size with default install. what are the consideration of the changes? is this a recommendation for user?

jdejoya17 replied:
there is no particular reason. this is the mount point of `filesystem` parameter. size is determined by the backing LV of the `filesystem` and not the size of the rootvg filesystem `/tmp`
however, i can make the mount point back to `/mnt`, i was just used to making the mount point that way during testing with filesystem

raeyacld replied:
We might want to keep the same for the shipped product if there is no errors or issues so we could avoid any unnecessary confusions to users

jdejoya17 replied:
i reverted back to `/mnt` for the example text

raeyacld commented on plugins/modules/mount.py:
```diff
@@ -171,10 +174,12 @@ def is_fspath_mounted(module, mount_dir, mount_over_dir):
-    if mount_dir:
-        cmd = "lsfs -l %s" % mount_dir
-    elif mount_over_dir:
+    if (mount_dir and mount_over_dir) or mount_over_dir:
```
can we simplify the check to "if mount_over_dir"? looks like no need to check mount_dir here

jdejoya17 replied:
you are correct. this is some leftover from previous logic. the order was different before.
i wanted to make sure that if both `mount_dir` and `mount_over_dir` was specified, i would
check for `mount_over_dir`, but since i changed the order to check for `mount_over_dir` first
then there is not need for that extra condition.
okay, i will simplify to
```
if mount_over_dir:
 ...
elif mount_dir:
 ...
else:
 ...
```

raeyacld commented on plugins/modules/mount.py:
```diff
@@ -250,13 +255,14 @@ def mount(module):
+        if is_fspath_mounted(module, mount_dir, mount_over_dir):
+            # if both mount_dir and mount_over_dir is given then check for
+            # mount_over_dir
+            if (mount_dir and mount_over_dir) or mount_over_dir:
```
Same here, should be checking mount_over_dir only sufficient?

jdejoya17 replied:
yes, you are right

raeyacld commented on plugins/modules/mount.py:
```diff
@@ -296,13 +302,14 @@ def umount(module):
+    if is_fspath_mounted(module, mount_dir, mount_over_dir) is False:
+        # if both mount_dir and mount_over_dir is given then check for
+        # mount_over_dir
+        if (mount_dir and mount_over_dir) or mount_over_dir:
```
Same here

jdejoya17 replied:
yes you are right

jdejoya17 committed:
- for the scenario where mount_dir and mount_over_dir are both given
then (1) mount_over_dir must be the one that should be check if it
is mounted; (2) the appropriate message would be used to report
mount_over_dir is already mounted/unmounted

jdejoya17 committed:
- fixed proper reporting when mounting/unmounting by mount
group, unmounting all remote filesystem, unmounting by node

jdejoya17 committed:
- added new unit test cases for mount module

jdejoya17 committed:
- added state action 'show', to list all mounted filesystems
- updated and added new examples for mount module
- fix typos, and simplified checking logic for mount_dir and
mount_over_dir

pvtorres commented on plugins/modules/mount.py:
```diff
@@ -105,24 +106,47 @@
-- name: List mounted filesystems
+- name: List mounted filesytems
   ibm.power_aix.mount:
+    state: list
```
Should this be state: show?

jdejoya17 replied:
yes you are correct. i will make changes regarding that

pvtorres commented on plugins/modules/mount.py:
```diff
@@ -154,6 +178,60 @@
 
+def is_mount_group_mounted(module, mount_group):
+    """
+    Determines which FS are already mounted in a monut group
```
mount => monut (spell)

jdejoya17 replied:
will make the changes

JLRebello commented on plugins/modules/mount.py:
```diff
@@ -105,24 +106,47 @@
 
 EXAMPLES = r'''
-- name: List mounted filesystems
+- name: List mounted filesytems
```
typo here - should be "filesystems"

jdejoya17 replied:
i will fix that.

JLRebello commented on plugins/modules/mount.py:
```diff
@@ -105,24 +106,47 @@
+    state: mount
+    mount_dir: /mnt/tesfs
 
 - name: Mount filesystems provided by a node"
```
typo - only 1 quotation mark here

jdejoya17 replied:
thanks, i will fix that
